esrally/mechanic/supplier.py

```diff
@@ -223,6 +231,58 @@ def __call__(self, *args, **kwargs):
         return binaries
 
 
+class CachedElasticsearchSourceSupplier:
+    def __init__(self, distributions_root, source_supplier, distribution_config, template_renderer):
+        self.distributions_root = distributions_root
+        self.source_supplier = source_supplier
+        self.template_renderer = template_renderer
+        self.cfg = distribution_config
+        self.runtime_jdk_bundled = convert.to_bool(self.cfg.get("runtime.jdk.bundled", False))
+        self.revision = None
+        self.cached_path = None
+        self.logger = logging.getLogger(__name__)
+
+    @property
+    def file_name(self):
+        if self.runtime_jdk_bundled:
+            url_key = "jdk.bundled.release_url"
+        else:
+            url_key = "jdk.unbundled.release_url"
+        url = self.template_renderer.render(self.cfg[url_key])
+        return url[url.rfind("/") + 1:]
+
+    @property
+    def cached(self):
+        return self.cached_path is not None and os.path.exists(self.cached_path)
+
+    def fetch(self):
+        resolved_revision = self.source_supplier.fetch()
+        if resolved_revision:
+            # ensure we use the resolved revision for rendering the artifact
+            self.template_renderer.version = resolved_revision
+            self.cached_path = os.path.join(self.distributions_root, self.file_name)
+
+    def prepare(self):
+        if not self.cached:
+            self.source_supplier.prepare()
+
+    def add(self, binaries):
+        if self.cached:
+            self.logger.info("Using cached artifact in [%s]", self.cached_path)
+            binaries["elasticsearch"] = self.cached_path
+        else:
+            self.source_supplier.add(binaries)
+            original_path = binaries["elasticsearch"]
+            # this can be None if the Elasticsearch does not reside in a git repo and the user has only
+            # copied all source files. In that case, we cannot resolve a revision hash and thus we cannot cache.
+            if self.cached_path:
+                shutil.copy(original_path, self.cached_path)
```

Comment:
What is there is some issue, out of disk space, file in use etc? Should we catch `IOError`, log it, and fallback to no caching?

Reply:
Thanks! I've pushed https://github.com/elastic/rally/pull/992/commits/937595eaa4661fe2fc2153f8b578a7e48cafb038.